src/routes/mint/[address]/[id].svelte

```diff
@@ -1,15 +1,141 @@
+	
+	let errorPreparingMint = false;
+	let readyToMint = false;
+	let newTokenURI = '';
+	
+	let newMetadata = {};
+	
 	onMount(async () => {
 		// Do initialization things here
+		
+		// const node1 = await Ipfs.create();
+		// const result = await node1.add("hello world2");
+		// debugger;
+		
+		
+		// If you directly load this page, the wallet isn't connected so it can't get blocknumber 
+		let blockNumber = "unknown";
+		if ($user.provider) {
+			blockNumber = await $user.provider.getBlockNumber();
+		}
+		
+		// First, get the token metadata (i.e. result of calling tokenURI and fetching)
+		const metadata = await getTokenMetadata();
+		newMetadata['originalTokenMetadata'] = metadata;
+		newMetadata['description'] = `This is a Snapshot of NFT ${metadata.name} at block #${blockNumber}`;
+		
+		// Then, retrieve the image so you can put it on ipfs
+		let image_url = metadata.image;
+		// returns 'ipfs', 'http', or 'embedded', or 'other'
+		const imageType = await getImageType(image_url);
+		console.log("image type:", imageType);
+		
 		const node = await Ipfs.create();
-		debugger;
-		// const client = await create('https://ipfs.infura.io:5001/api/v0');
-		// let results = await client.add('a string that will live on IPFS');
-		console.log(results);
+		
+		// TODO: eliminate false positives due to http://gateway/ipfs
+		// They should be treated as IPFS links and may not need to be re-added to ipfs
+		if (imageType == 'http') {
+			// Get image data 
+			const response = await fetch(image_url);
+			if (!response.ok) {  // TODO: Show fail to fetch in UI 
+				console.log('Issue fetching image: ', response);
+				errorPreparingMint = true;
+				return;
+			}
+			let result;
+			try {
+				const imageData = await response.blob();
+				result = await node.add(imageData);
+				console.log("ipfs result", result);
+				
+				// if you get ipfs, overwrite the image_url
+				image_url = "ipfs://" + result.path + "/";
+			} catch (error) {
+				console.log('Error uploading image to IPFS: ', error);
+				errorPreparingMint = true;
+				return; 
+			}
+		}
+		
+		newMetadata['image_url'] = image_url;
+		newMetadata['image'] = getPolaroidVersion(image_url);
+		
+		// Now newMetadata is complete, upload it so we can use it as the tokenURI
+		try {
+			const result = await node.add(JSON.stringify(newMetadata));
+			console.log("metadata ipfs result", result);
+			
+			// if you get ipfs, overwrite the image_url
+			newTokenURI = "ipfs://" + result.path + "/";
+		} catch (error) {
+			console.log('Error uploading metadata to IPFS: ', error);
+			errorPreparingMint = true;
+			return; 
+		}
+		
+		readyToMint = true;
+		mint();
 	});
+	
+	async function mint() {
+		// TODO: mint with newTokenURI
+	}
+	
+	async function getTokenMetadata() {
+		// Note(divergio): I could not get this code that reads the contract to work
+		// let nftContract: Contract = new ethers.Contract($page.params.address, defaultAbi, $user.provider);
```

Comment:
This code: did anyone get this working for arbitrary NFT contracts?? @johnmosesman I think you were working on this last week.

Reply:
Yes we had a version of this earlier/still in the project - https://github.com/ryanlabouve/nftrusty-exploration/commit/e030f055c2a2385c9ad43cc69342b1fa3a7a0c96#diff-5904556f9755a5387a4ebe2d4e9059da6e5704606ca0ce2d3f6d2482087305dcR92

Basically we put an ABI in the project that responds to TokenURI, and we are assuming that the contract you're calling will respond to it.

Not saying this is the best way or that your Moralis way isn't better, just what we had.

Reply:
It wasn't working for me last night, but I might have just had some typo or something. 

I strongly prefer we query the contract. The reason being that Moralis's data is outdated. It defeats the purpose of the snapshot, because the NFT might have changed since Moralis scraped it.

Reply:
The advantage of Moralis may be that they handle more edge cases than our simple method does